activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1042,7 +1042,11 @@ def build_select(arel)
         if select_values.any?
           arel.project(*arel_columns(select_values.uniq))
         else
-          arel.project(@klass.arel_table[Arel.star])
+          if @klass.ignored_columns.any?
+            arel.project(*arel_columns(@klass.columns.map(&:name)))
```

Comment:
There's a [memoized `column_names` method](https://github.com/rails/rails/blob/v5.0.1/activerecord/lib/active_record/model_schema.rb#L384-L386) you could use here.